aspects/aspects.go

```diff
@@ -527,6 +586,195 @@ func NewJSONDataBag() JSONDataBag {
 	return JSONDataBag(make(map[string]json.RawMessage))
 }
 
+func (s JSONDataBag) Query(path string, params map[string]string, filters map[string]Filter) ([]interface{}, error) {
+	subKeys := splitPath(path)
+	return query(subKeys, 0, s, params, filters)
+}
+
+func query(subKeys []string, index int, node map[string]json.RawMessage, params map[string]string, filters map[string]Filter) ([]interface{}, error) {
+	key := subKeys[index]
+
+	var fieldFilter filter
+	if key[0] == '{' {
+		var err error
+		key, fieldFilter, err = parsePlaceholder(key, params, filters)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var rawLevels []json.RawMessage
+	if key == "*" {
+		// if there wasn't a mapping for the placeholder, read all
+		for _, val := range node {
+			rawLevels = append(rawLevels, val)
+		}
+	} else {
+		nextLevel, ok := node[key]
+		if !ok {
+			pathPrefix := strings.Join(subKeys[:index+1], ".")
+			return nil, PathNotFoundError(fmt.Sprintf("no value was found under path %q", pathPrefix))
+		}
+
+		rawLevels = []json.RawMessage{nextLevel}
+	}
+
+	// read the final value
+	if index == len(subKeys)-1 {
+		var results []interface{}
+
+		for _, val := range rawLevels {
+			var res interface{}
+
+			// if there's a field filter, assume it's a map
+			if fieldFilter != nil {
+				var nextLevel map[string]json.RawMessage
+				if err := json.Unmarshal(val, &nextLevel); err != nil {
+					var typeErr *json.UnmarshalTypeError
+					if errors.As(err, &typeErr) {
+						return nil, fmt.Errorf("cannot filter by field: expected a map but got a %s", typeErr.Value)
+					}
+					return nil, err
+				}
+
+				pass, err := fieldFilter(nextLevel)
+				if err != nil {
+					return nil, err
+				}
+
+				if !pass {
+					continue
+				}
+				res = nextLevel
+			} else {
+				if err := json.Unmarshal(val, &res); err != nil {
+					var typeErr *json.UnmarshalTypeError
+					if errors.As(err, &typeErr) {
+						path := strings.Join(subKeys, ".")
+						return nil, fmt.Errorf("cannot read value of %q into %T: maps to %s", path, res, typeErr.Value)
+					}
+
+					return nil, err
+				}
+			}
+
+			results = append(results, res)
```

Comment:
the shape of results is one of the open questions we have, in principle it would follow the structure of the request nested namespace